File: packages/verifier-py/src/primust_verify/verifier.py
"""primust-verify — Offline VPEC artifact verifier.

ZERO runtime dependencies on Primust infrastructure after initial
public key fetch. Must verify a VPEC produced today in 10 years.
"""

from __future__ import annotations

import base64
import hashlib
import json
import re
from typing import Any, Optional

from primust_artifact_core.validate_artifact import validate_artifact
from primust_artifact_core.signing import verify as ed25519_verify
from primust_artifact_core.types import SignatureEnvelope

from primust_verify.key_cache import get_key
from primust_verify.types import VerifyOptions, VerificationResult


def _has_reliance_mode(obj: dict[str, Any], path: str = "") -> Optional[str]:
    """Recursively check for reliance_mode field anywhere."""
    for key, value in obj.items():
        current_path = f"{path}.{key}" if path else key
        if key == "reliance_mode":
            return current_path
        if isinstance(value, dict):
            found = _has_reliance_mode(value, current_path)
            if found:
                return found
    return None


def _base_result(artifact: dict[str, Any]) -> VerificationResult:
    """Build a default VerificationResult from an artifact dict."""
    sig = artifact.get("signature") or {}
    issuer = artifact.get("issuer") or {}
    proof_dist = artifact.get("proof_distribution") or {}
    coverage = artifact.get("coverage") or {}
    gaps_raw = artifact.get("gaps", [])

    gaps = []
    if isinstance(gaps_raw, list):
        for g in gaps_raw:
            if isinstance(g, dict):
                gaps.append({
                    "gap_id": g.get("gap_id", ""),
                    "gap_type": g.get("gap_type", ""),
                    "severity": g.get("severity", ""),
                })

    return VerificationResult(
        vpec_id=artifact.get("vpec_id", ""),
        valid=False,
        schema_version=artifact.get("schema_version", ""),
        proof_level=artifact.get("proof_level", ""),
        proof_distribution=proof_dist if isinstance(proof_dist, dict) else {},
        org_id=artifact.get("org_id", ""),
        workflow_id=artifact.get("workflow_id", ""),
        process_context_hash=artifact.get("process_context_hash"),
        partial=artifact.get("partial", False),
        test_mode=artifact.get("test_mode", False),
        signer_id=issuer.get("signer_id", "") or sig.get("signer_id", ""),
        kid=issuer.get("kid", "") or sig.get("kid", ""),
        signed_at=sig.get("signed_at", ""),
        timestamp_anchor_valid=None,
        rekor_status="skipped",
        zk_proof_valid=None,
        manifest_hashes={},
        gaps=gaps,
        coverage=coverage if isinstance(coverage, dict) else {},
        errors=[],
        warnings=[],
    )


def _extract_key_from_pem(pem: str) -> str:
    """Extract base64url key from PEM or raw base64url string."""
    if "-----BEGIN" in pem:
        b64 = re.sub(r"-----[A-Z ]+-----", "", pem).strip()
        b64 = b64.replace("\n", "").replace("\r", "")
        return b64.replace("+", "-").replace("/", "_").rstrip("=")
    return pem.strip()


def verify(
    artifact: dict[str, Any],
    options: Optional[VerifyOptions] = None,
) -> VerificationResult:
    """Verify a VPEC artifact.

    Args:
        artifact: Parsed artifact JSON dict.
        options: Verification options.

    Returns:
        VerificationResult with errors/warnings.
    """
    if options is None:
        options = VerifyOptions()

    result = _base_result(artifact)

    # ── Step 1: Schema validation ──
    schema_result = validate_artifact(artifact)
    if not schema_result.valid:
        for err in schema_result.errors:
            if err.code == "RELIANCE_MODE_FORBIDDEN":
                result.errors.append("banned_field_reliance_mode")
            elif err.code == "MANIFEST_HASHES_NOT_MAP":
                result.errors.append("manifest_hashes_not_object")
            else:
                result.errors.append(f"schema_validation_failed: {err.code}")
        return result

    # Extra check: reliance_mode anywhere
    reliance_path = _has_reliance_mode(artifact)
    if reliance_path:
        result.errors.append("banned_field_reliance_mode")
        return result

    # ── Step 4 (early): Kid resolution ──
    issuer = artifact.get("issuer", {})
    sig = artifact.get("signature", {})

    if not issuer or not sig:
        result.errors.append("missing_issuer_or_signature")
        return result

    if issuer.get("kid") != sig.get("kid"):
        result.errors.append("kid_mismatch")
        return result

    # ── Step 2+3: Integrity + Ed25519 signature ──
    document_body = {k: v for k, v in artifact.items() if k != "signature"}

    # Resolve public key
    try:
        pem = get_key(
            sig["kid"],
            issuer.get("public_key_url", ""),
            options.trust_root,
        )
        public_key_b64url = _extract_key_from_pem(pem)
    except Exception as e:
        result.errors.append(str(e))
        return result

    signature_envelope = SignatureEnvelope(
        signer_id=sig.get("signer_id", ""),
        kid=sig.get("kid", ""),
        algorithm=sig.get("algorithm", "Ed25519"),
        signature=sig.get("signature", ""),
        signed_at=sig.get("signed_at", ""),
    )

    sig_valid = ed25519_verify(document_body, signature_envelope, public_key_b64url)
    if not sig_valid:
        result.errors.append("integrity_check_failed")
        return result

    # ── Step 5: Signer status check (Rekor) ──
    if options.skip_network:
        result.rekor_status = "skipped"
    else:
        result.rekor_status = _check_rekor(public_key_b64url, sig.get("kid", ""))
        if result.rekor_status == "unavailable":
            result.warnings.append("rekor_check_unavailable")
        elif result.rekor_status == "revoked":
            result.errors.append("signer_key_revoked")
            return result

    # ── Step 6: RFC 3161 timestamp verification ──
    ts_anchor = artifact.get("timestamp_anchor")
    if isinstance(ts_anchor, dict) and ts_anchor.get("type") == "rfc3161" and ts_anchor.get("value"):
        result.timestamp_anchor_valid = _verify_timestamp_imprint(
            ts_anchor["value"], document_body,
        )
        if result.timestamp_anchor_valid is False:
            result.warnings.append("rfc3161_imprint_mismatch")
        elif result.timestamp_anchor_valid is True:
            result.warnings.append("rfc3161_tsa_cert_chain_not_verified")
    else:
        result.timestamp_anchor_valid = None

    # ── Step 7: Proof level integrity ──
    proof_dist = artifact.get("proof_distribution", {})
    if artifact.get("proof_level") != proof_dist.get("weakest_link"):
        result.errors.append("proof_level_mismatch")
        return result

    # ── Step 8: Manifest hash audit ──
    manifest_hashes = artifact.get("manifest_hashes", {})
    if isinstance(manifest_hashes, dict):
        result.manifest_hashes = manifest_hashes

    # ── Step 9: ZK proof verification ──
    pending_flags = artifact.get("pending_flags", {})
    if artifact.get("zk_proof") and not pending_flags.get("proof_pending"):
        zk_proof = artifact["zk_proof"]
        proving_system = zk_proof.get("proving_system") if isinstance(zk_proof, dict) else None

        if proving_system == "ultrahonk":
            result.zk_proof_valid = _verify_ultrahonk(zk_proof)
            if result.zk_proof_valid is False:
                result.errors.append("zk_proof_invalid")
        elif proving_system == "ezkl":
            # EZKL Tier 2: explicit stub — requires EZKL verifier integration
            result.zk_proof_valid = None
            result.warnings.append("ezkl_verification_not_implemented")
        else:
            result.zk_proof_valid = None
            result.warnings.append(f"unknown_proving_system: {proving_system or 'none'}")
    else:
        result.zk_proof_valid = None

    # ── Step 9b: Mathematical proof_level requires verified ZK proof ──
    if artifact.get("proof_level") == "mathematical" and result.zk_proof_valid is None:
        result.errors.append("mathematical_proof_not_verified")
        return result

    # ── Step 10: test_mode check ──
    if artifact.get("test_mode") is True:
        if options.production:
            result.errors.append("test_mode_rejected_in_production")
            return result
        result.warnings.append("test_credential")

    # All checks passed
    result.valid = len(result.errors) == 0
    return result


# ── RFC 3161 Timestamp Imprint Verification ──


def _verify_timestamp_imprint(
    ts_token_b64: str,
    document_body: dict[str, Any],
) -> bool | None:
    """Verify the message imprint inside an RFC 3161 TimeStampResp matches
    SHA-256(canonical(documentBody)).

    Parses enough DER to extract the hashed message from the MessageImprint field.
    Returns True if imprint matches, False if mismatch, None if unparseable.
    """
    try:
        from primust_artifact_core.canonical import canonical
    except ImportError:
        return None

    try:
        ts_resp = base64.b64decode(ts_token_b64)

        # Find SHA-256 OID (2.16.840.1.101.3.4.2.1) in the DER
        sha256_oid = bytes([0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01])
        oid_idx = ts_resp.find(sha256_oid)
        if oid_idx == -1:
            return None

        # After OID + NULL, look for OCTET STRING (0x04) containing 32-byte hash
        search_start = oid_idx + len(sha256_oid)
        for i in range(search_start, min(search_start + 20, len(ts_resp) - 33)):
            if ts_resp[i] == 0x04 and ts_resp[i + 1] == 0x20:
                extracted_hash = ts_resp[i + 2 : i + 2 + 32]

                # Recompute expected hash
                canonical_doc = canonical(document_body)
                expected_hash = hashlib.sha256(canonical_doc.encode()).digest()

                return extracted_hash == expected_hash

        return None  # Could not find hash in DER
    except Exception:
        return None


# ── Rekor Status Check ──

REKOR_API = "https://rekor.sigstore.dev/api/v1"


def _check_rekor(public_key_b64url: str, kid: str) -> str:
    """Check Rekor for key revocation by querying with SHA-256 fingerprint
    of the public key bytes.

    Returns: 'active', 'not_found', 'revoked', or 'unavailable'.
    """
    try:
        import urllib.request

        # Decode public key bytes and compute SHA-256 fingerprint
        # base64url → base64 → bytes
        b64 = public_key_b64url.replace("-", "+").replace("_", "/")
        padding = 4 - len(b64) % 4
        if padding != 4:
            b64 += "=" * padding
        key_bytes = base64.b64decode(b64)
        fingerprint = hashlib.sha256(key_bytes).hexdigest()

        # Search Rekor index by key fingerprint
        req_body = json.dumps({"hash": f"sha256:{fingerprint}"}).encode()
        req = urllib.request.Request(
            f"{REKOR_API}/index/retrieve",
            data=req_body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=5) as resp:
            entries = json.loads(resp.read())

        if not entries:
            return "not_found"

        return "active"
    except Exception:
        return "unavailable"


# ── ZK Proof Verification (UltraHonk) ──


def _verify_ultrahonk(zk_proof: dict[str, Any]) -> bool | None:
    """Verify an UltraHonk ZK proof.

    Requires the `bb` CLI tool (Barretenberg) to be installed.
    Returns True if valid, False if invalid, None if verification unavailable.
    """
    try:
        import subprocess
        import tempfile

        proof_b64 = zk_proof.get("proof")
        vk_b64 = zk_proof.get("verification_key")

        if not proof_b64 or not vk_b64:
            return False

        proof_bytes = base64.b64decode(proof_b64)
        vk_bytes = base64.b64decode(vk_b64)

        with tempfile.NamedTemporaryFile(suffix=".proof", delete=False) as pf:
            pf.write(proof_bytes)
            proof_path = pf.name

        with tempfile.NamedTemporaryFile(suffix=".vk", delete=False) as vf:
            vf.write(vk_bytes)
            vk_path = vf.name

        # Use bb CLI for verification
        result = subprocess.run(
            ["bb", "verify_ultra_honk", "-p", proof_path, "-k", vk_path],
            capture_output=True,
            timeout=30,
        )

        import os
        os.unlink(proof_path)
        os.unlink(vk_path)

        return result.returncode == 0
    except FileNotFoundError:
        # bb CLI not installed
        return None
    except Exception:
        return None

File: packages/verifier/src/verifier.ts
/**
 * primust-verify — Offline VPEC artifact verifier.
 *
 * ZERO runtime dependencies on Primust infrastructure after initial
 * public key fetch. Must verify a VPEC produced today in 10 years.
 *
 * Verification steps (in order):
 * 1. Schema validation
 * 2. SHA-256 integrity + Ed25519 signature (combined)
 * 3. Kid resolution
 * 4. Signer status check (Rekor — stubbed in v1)
 * 5. RFC 3161 timestamp verification (stubbed in v1)
 * 6. Proof level integrity
 * 7. Manifest hash audit
 * 8. ZK proof verification (stubbed in v1)
 * 9. test_mode check
 */

import { createHash } from 'node:crypto';
import {
  canonical,
  verify as ed25519Verify,
  validateArtifact,
  fromBase64Url,
} from '@primust/artifact-core';
import type { SignatureEnvelope } from '@primust/artifact-core';
import { getKey } from './key-cache.js';
import type { VerifyOptions, VerificationResult, RekorStatus } from './types.js';

/**
 * Check recursively for reliance_mode field anywhere in the artifact.
 */
function hasRelianceMode(obj: Record<string, unknown>, path = ''): string | null {
  for (const [key, value] of Object.entries(obj)) {
    const currentPath = path ? `${path}.${key}` : key;
    if (key === 'reliance_mode') {
      return currentPath;
    }
    if (value && typeof value === 'object' && !Array.isArray(value)) {
      const found = hasRelianceMode(value as Record<string, unknown>, currentPath);
      if (found) return found;
    }
  }
  return null;
}

/**
 * Build a default (failed/empty) VerificationResult from an artifact dict.
 */
function baseResult(artifact: Record<string, unknown>): VerificationResult {
  const sig = artifact.signature as Record<string, unknown> | undefined;
  const issuer = artifact.issuer as Record<string, unknown> | undefined;
  const proofDist = artifact.proof_distribution as Record<string, unknown> | undefined;
  const coverage = artifact.coverage as Record<string, unknown> | undefined;
  const gaps = Array.isArray(artifact.gaps) ? artifact.gaps : [];

  return {
    vpec_id: (artifact.vpec_id as string) ?? '',
    valid: false,
    schema_version: (artifact.schema_version as string) ?? '',
    proof_level: (artifact.proof_level as string) ?? '',
    proof_distribution: proofDist ?? {},
    org_id: (artifact.org_id as string) ?? '',
    workflow_id: (artifact.workflow_id as string) ?? '',
    process_context_hash: (artifact.process_context_hash as string | null) ?? null,
    partial: (artifact.partial as boolean) ?? false,
    test_mode: (artifact.test_mode as boolean) ?? false,
    signer_id: (issuer?.signer_id as string) ?? (sig?.signer_id as string) ?? '',
    kid: (issuer?.kid as string) ?? (sig?.kid as string) ?? '',
    signed_at: (sig?.signed_at as string) ?? '',
    timestamp_anchor_valid: null,
    rekor_status: 'skipped',
    zk_proof_valid: null,
    manifest_hashes: {},
    gaps: gaps.map((g: Record<string, unknown>) => ({
      gap_id: (g.gap_id as string) ?? '',
      gap_type: (g.gap_type as string) ?? '',
      severity: (g.severity as string) ?? '',
    })),
    coverage: coverage ?? {},
    errors: [],
    warnings: [],
  };
}

/**
 * Verify a VPEC artifact.
 *
 * @param artifact - Parsed artifact JSON (Record<string, unknown>)
 * @param options  - Verification options
 * @returns VerificationResult with errors/warnings
 */
export async function verify(
  artifact: Record<string, unknown>,
  options: VerifyOptions = {},
): Promise<VerificationResult> {
  const result = baseResult(artifact);
  const { production = false, skip_network = false, trust_root } = options;

  // ── Step 1: Schema validation ──
  const schemaResult = validateArtifact(artifact);
  if (!schemaResult.valid) {
    for (const err of schemaResult.errors) {
      if (err.code === 'RELIANCE_MODE_FORBIDDEN') {
        result.errors.push('banned_field_reliance_mode');
      } else if (err.code === 'MANIFEST_HASHES_NOT_MAP') {
        result.errors.push('manifest_hashes_not_object');
      } else {
        result.errors.push(`schema_validation_failed: ${err.code}`);
      }
    }
    return result;
  }

  // Extra check: reliance_mode anywhere (validateArtifact already catches this,
  // but we ensure the specific error string)
  const reliancePath = hasRelianceMode(artifact);
  if (reliancePath) {
    result.errors.push('banned_field_reliance_mode');
    return result;
  }

  // ── Step 4 (early): Kid resolution — issuer.kid must match signature.kid ──
  const issuer = artifact.issuer as Record<string, unknown>;
  const sig = artifact.signature as Record<string, unknown>;

  if (!issuer || !sig) {
    result.errors.push('missing_issuer_or_signature');
    return result;
  }

  if (issuer.kid !== sig.kid) {
    result.errors.push('kid_mismatch');
    return result;
  }

  // ── Step 2+3: Integrity + Ed25519 signature verification ──
  // The signing process signs the artifact body (everything except 'signature').
  // We reconstruct the document by stripping the signature field.
  const { signature: _sig, ...documentBody } = artifact;

  // Resolve public key
  let publicKeyB64Url: string;
  try {
    const pem = await getKey(
      sig.kid as string,
      issuer.public_key_url as string,
      trust_root,
    );
    // PEM may be raw base64url or PEM-wrapped. Handle both.
    publicKeyB64Url = extractKeyFromPem(pem);
  } catch (err) {
    result.errors.push((err as Error).message);
    return result;
  }

  const signatureEnvelope: SignatureEnvelope = {
    signer_id: sig.signer_id as string,
    kid: sig.kid as string,
    algorithm: sig.algorithm as 'Ed25519',
    signature: sig.signature as string,
    signed_at: sig.signed_at as string,
  };

  const sigValid = ed25519Verify(documentBody, signatureEnvelope, publicKeyB64Url);
  if (!sigValid) {
    result.errors.push('integrity_check_failed');
    return result;
  }

  // ── Step 5: Signer status check (Rekor) ──
  if (skip_network) {
    result.rekor_status = 'skipped';
  } else {
    result.rekor_status = await checkRekor(publicKeyB64Url, sig.kid as string);
    if (result.rekor_status === 'unavailable') {
      result.warnings.push('rekor_check_unavailable');
    } else if (result.rekor_status === 'revoked') {
      result.errors.push('signer_key_revoked');
      return result;
    }
  }

  // ── Step 6: RFC 3161 timestamp verification ──
  const tsAnchor = artifact.timestamp_anchor as Record<string, unknown> | undefined;
  if (tsAnchor && tsAnchor.type === 'rfc3161' && typeof tsAnchor.value === 'string') {
    result.timestamp_anchor_valid = verifyTimestampImprint(
      tsAnchor.value as string,
      documentBody,
    );
    if (result.timestamp_anchor_valid === false) {
      result.warnings.push('rfc3161_imprint_mismatch');
    } else if (result.timestamp_anchor_valid === true) {
      result.warnings.push('rfc3161_tsa_cert_chain_not_verified');
    }
  } else {
    result.timestamp_anchor_valid = null;
  }

  // ── Step 7: Proof level integrity ──
  // Already checked by validateArtifact in step 1 (PROOF_LEVEL_MISMATCH),
  // but we verify again explicitly for the spec requirement.
  const proofDist = artifact.proof_distribution as Record<string, unknown>;
  if (artifact.proof_level !== proofDist.weakest_link) {
    result.errors.push('proof_level_mismatch');
    return result;
  }

  // ── Step 8: Manifest hash audit ──
  const manifestHashes = artifact.manifest_hashes as Record<string, string>;
  result.manifest_hashes = manifestHashes;

  // ── Step 9: ZK proof verification ──
  const pendingFlags = artifact.pending_flags as Record<string, unknown> | undefined;
  const proofPending = pendingFlags?.proof_pending === true;

  if (artifact.zk_proof && !proofPending) {
    const zkProof = artifact.zk_proof as Record<string, unknown>;
    // Accept both 'prover_system' (canonical) and 'proving_system' (legacy)
    const provingSystem = (zkProof.prover_system ?? zkProof.proving_system) as string | undefined;

    if (provingSystem === 'ultrahonk') {
      result.zk_proof_valid = await verifyUltraHonk(zkProof);
      if (result.zk_proof_valid === false) {
        result.errors.push('zk_proof_invalid');
      }
    } else if (provingSystem === 'ezkl') {
      // EZKL Tier 2: explicit stub — requires EZKL verifier integration
      result.zk_proof_valid = null;
      result.warnings.push('ezkl_verification_not_implemented');
    } else {
      result.zk_proof_valid = null;
      result.warnings.push(`unknown_proving_system: ${provingSystem ?? 'none'}`);
    }
  } else if (proofPending) {
    // Proof in flight — cannot verify yet
    result.zk_proof_valid = null;
    result.warnings.push('proof_pending');
  } else {
    result.zk_proof_valid = null;
  }

  // ── Step 9b: Mathematical proof_level requires verified ZK proof ──
  if (artifact.proof_level === 'mathematical') {
    if (proofPending) {
      // proof_pending + mathematical: proof hasn't arrived yet — warning, not error
      result.warnings.push('mathematical_proof_pending');
    } else if (result.zk_proof_valid === null) {
      // No proof and not pending — this is an error
      result.errors.push('mathematical_proof_not_verified');
      return result;
    }
  }

  // ── Step 10: test_mode check ──
  if (artifact.test_mode === true) {
    if (production) {
      result.errors.push('test_mode_rejected_in_production');
      return result;
    }
    result.warnings.push('test_credential');
  }

  // All checks passed
  result.valid = result.errors.length === 0;
  return result;
}

/**
 * Extract base64url key from PEM or raw base64url string.
 * Handles both PEM-wrapped keys and raw base64url.
 */
function extractKeyFromPem(pem: string): string {
  // If it looks like PEM, strip header/footer and convert base64 → base64url
  if (pem.includes('-----BEGIN')) {
    const b64 = pem
      .replace(/-----BEGIN [A-Z ]+-----/g, '')
      .replace(/-----END [A-Z ]+-----/g, '')
      .replace(/\s/g, '');
    // Convert standard base64 to base64url
    return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  }
  // Already base64url
  return pem.trim();
}

// ── RFC 3161 Timestamp Imprint Verification ──

/**
 * Verify the message imprint inside an RFC 3161 TimeStampResp matches
 * SHA-256(canonical(documentBody)).
 *
 * Parses enough DER to extract the hashed message from the MessageImprint
 * field. Returns true if imprint matches, false if mismatch, null if unparseable.
 */
function verifyTimestampImprint(
  tsTokenB64: string,
  documentBody: Record<string, unknown>,
): boolean | null {
  try {
    const tsResp = Buffer.from(tsTokenB64, 'base64');

    // Find SHA-256 OID (2.16.840.1.101.3.4.2.1) in the DER
    const sha256Oid = Buffer.from([0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01]);
    const oidIdx = findBuffer(tsResp, sha256Oid);
    if (oidIdx === -1) return null;

    // The message imprint hash follows the AlgorithmIdentifier.
    // After OID + NULL, look for OCTET STRING (0x04) containing 32-byte hash.
    const searchStart = oidIdx + sha256Oid.length;
    for (let i = searchStart; i < Math.min(searchStart + 20, tsResp.length - 33); i++) {
      if (tsResp[i] === 0x04 && tsResp[i + 1] === 0x20) {
        const extractedHash = tsResp.subarray(i + 2, i + 2 + 32);

        // Recompute expected hash
        const canonicalDoc = canonical(documentBody);
        const expectedHash = createHash('sha256').update(canonicalDoc).digest();

        return extractedHash.equals(expectedHash);
      }
    }
    return null; // Could not find hash in DER
  } catch {
    return null;
  }
}

function findBuffer(haystack: Buffer, needle: Buffer): number {
  for (let i = 0; i <= haystack.length - needle.length; i++) {
    if (haystack.subarray(i, i + needle.length).equals(needle)) return i;
  }
  return -1;
}

// ── Rekor Status Check ──

const REKOR_API = 'https://rekor.sigstore.dev/api/v1';

/**
 * Check Rekor for key revocation by querying with SHA-256 fingerprint
 * of the public key bytes.
 */
async function checkRekor(publicKeyB64Url: string, kid: string): Promise<RekorStatus> {
  try {
    // Decode public key bytes and compute SHA-256 fingerprint
    const keyBytes = fromBase64Url(publicKeyB64Url);
    const fingerprint = createHash('sha256').update(Buffer.from(keyBytes)).digest('hex');

    // Search Rekor index by key fingerprint
    const resp = await fetch(`${REKOR_API}/index/retrieve`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ hash: `sha256:${fingerprint}` }),
      signal: AbortSignal.timeout(5000),
    });

    if (!resp.ok) {
      return 'unavailable';
    }

    const entries = await resp.json() as string[];
    if (!entries || entries.length === 0) {
      // No entries found — key has not been submitted to Rekor (not necessarily bad)
      return 'not_found';
    }

    // Key found in Rekor — it's been logged (active, not revoked)
    return 'active';
  } catch {
    return 'unavailable';
  }
}

// ── ZK Proof Verification (UltraHonk) ──

/**
 * Verify an UltraHonk ZK proof using @aztec/bb.js.
 *
 * Requires:
 * - zk_proof.proof: base64-encoded proof bytes
 * - zk_proof.public_inputs: array of field elements (hex strings)
 * - zk_proof.verification_key: base64-encoded verification key
 *
 * Returns true if valid, false if invalid, null if verification unavailable.
 */
async function verifyUltraHonk(zkProof: Record<string, unknown>): Promise<boolean | null> {
  try {
    // Dynamic import — @aztec/bb.js is an optional dependency
    const bb = await import('@aztec/bb.js').catch(() => null);
    if (!bb) {
      return null; // bb.js not available
    }

    const proofB64 = zkProof.proof as string;
    const publicInputs = zkProof.public_inputs as string[];
    const vkB64 = zkProof.verification_key as string;

    if (!proofB64 || !publicInputs || !vkB64) {
      return false;
    }

    const proofBytes = Buffer.from(proofB64, 'base64');
    const vkBytes = Buffer.from(vkB64, 'base64');

    // UltraHonk verification
    const api = await bb.newBarretenbergApiAsync();
    const valid = await api.acirVerifyUltraHonk(proofBytes, vkBytes);
    return valid;
  } catch {
    return null; // Verification error — treat as unavailable, not invalid
  }
}
